spec/02-integration/09-hybrid_mode/12-errors_spec.lua

```diff
@@ -0,0 +1,285 @@
+
+local CLUSTER_PORT = helpers.get_available_port()
+local CLUSTER_SSL_PORT = helpers.get_available_port()
+
+
+local function json(data)
+  return {
+    headers = {
+      ["accept"] = "application/json",
+      ["content-type"] = "application/json",
+    },
+    body = assert(cjson.encode(data)),
+  }
+end
+
+
+local mock_cp = ([[
+  lua_shared_dict kong_test_cp_mock 10m;
+
+  server {
+      charset UTF-8;
+      server_name kong_cluster_listener;
+      listen %s;
+      listen %s ssl;
+
+      access_log ${{ADMIN_ACCESS_LOG}};
+      error_log  ${{ADMIN_ERROR_LOG}} ${{LOG_LEVEL}};
+
+> if cluster_mtls == "shared" then
+      ssl_verify_client   optional_no_ca;
+> else
+      ssl_verify_client   on;
+      ssl_client_certificate ${{CLUSTER_CA_CERT}};
+      ssl_verify_depth     4;
+> end
+      ssl_certificate     ${{CLUSTER_CERT}};
+      ssl_certificate_key ${{CLUSTER_CERT_KEY}};
+      ssl_session_cache   shared:ClusterSSL:10m;
+
+      location = /v1/outlet {
+          content_by_lua_block {
+              require("spec.fixtures.mock_cp").outlet()
+          }
+      }
+
+      location = /payload {
+          content_by_lua_block {
+              require("spec.fixtures.mock_cp").set_payload()
+          }
+      }
+
+      location = /log {
+          content_by_lua_block {
+              require("spec.fixtures.mock_cp").get_log()
+          }
+      }
+  }
+]]):format(CLUSTER_PORT, CLUSTER_SSL_PORT)
+
+
+local function set_cp_payload(client, payload)
+  local res = client:post("/payload", json(payload))
+  assert.response(res).has.status(201)
+end
+
+
+local function get_connection_log(client)
+  local res = client:get("/log")
+  assert.response(res).has.status(200)
+  local body = assert.response(res).has.jsonbody()
+  assert.is_table(body.data)
+
+  return body.data
+end
+
+
+---@param client table
+---@param msg string
+---@return { error: kong.clustering.config_helper.update.err_t }
+local function get_error_report(client, msg)
+  local err_t
+
+  assert.eventually(function()
+    local entries = get_connection_log(client)
+
+    if #entries == 0 then
+      return nil, { err = "no data plane client log entries" }
+    end
+
+    for _, entry in ipairs(entries) do
+      if    entry.event == "client-recv"
+        and entry.type  == "binary"
+        and type(entry.json) == "table"
+        and entry.json.type == "error"
+        then
+          err_t = entry.json
+          return true
+        end
+      end
+
+      return nil, {
+        err = "did not find expected error in log",
+        entries = entries,
+      }
+    end)
+  .is_truthy(msg)
+
+  return err_t
+end
+
+
+for _, strategy in helpers.each_strategy() do
+  describe("CP/DP sync error-reporting with #" .. strategy .. " backend", function()
+    local client
+    local fixtures = { http_mock = { control_plane = mock_cp } }
+
+    before_each(function()
+      helpers.clean_prefix()
+
+      assert(helpers.start_kong({
```

Comment:
Is it relevant to start and stop the gateway on every test? Could we use `lazy_setup` and `lazy_teardown` to speed up the tests?

Reply:
My assumption was that the tests would be more consistent if we started fresh each time, but I don't think I've tested this idea. I'll try it out and see if I can avoid the start/stop each time.

Reply:
Update: I tried this, it works, and the test runs in ~4.5s instead of ~7s! :trophy: 

https://i.giphy.com/14nEjMEbnUkFvq.webp